File: src/Parsers/APIParser.cpp
#include "Parsers/APIParser.h"

#include <iostream>

using std::cout;
using std::endl;

using Models::API;
using Models::IcarusAction;

namespace Parsers
{

#pragma region Constructors
APIParser::APIParser(IcarusAction icaAct) : icaAct(icaAct)
{
    api = API{};
    parseAPI();
}
#pragma endregion


#pragma region Functions
API APIParser::retrieveAPI() const
{
    return api;
}

void APIParser::parseAPI()
{
    auto flags = icaAct.flags;
    cout << "Parsing api" << endl;

    for (auto i =0; i < flags.size(); ++i)
    {
        auto arg = flags[i].flag;
        auto value = flags[i].value;

        if (arg.compare("-h") == 0)
        {
            api.url = (value[value.size()-1] == '/') ? value : value + "/";
            break;
        }
    }

    // NOTE: For now I will hard code
    // the api version since I am only
    // on version 1
    api.version = "v1";
}
#pragma endregion

}

File: src/parsers/api_parser.rs
use crate::models;

#[derive(Clone, Debug)]
pub struct APIParser {
    pub apis: Vec<models::api::Api>,
    pub ica_act: models::icarus_action::IcarusAction,
}

pub const API_VERSION: &str = "v2";

pub enum APIType {
    Main,
    Auth,
}

impl APIParser {
    pub fn retrieve_api(&self, api_type: APIType) -> models::api::Api {
        match api_type {
            APIType::Main => self.apis[0].clone(),
            APIType::Auth => self.apis[1].clone(),
        }
    }

    pub fn parse_api(&mut self, api_type: APIType) {
        let flags = self.ica_act.flags.clone();
        println!("Parsing api");

        for elem in flags {
            let arg = elem.flag;
            let value = elem.value;

            match api_type {
                APIType::Main => {
                    if arg == "-h" {
                        if value.chars().nth(value.len() - 1) == Some('/') {
                            self.apis[0].url = value;
                        } else {
                            self.apis[0].url = value + "/";
                        }
                        break;
                    }
                }
                APIType::Auth => {
                    if arg == "-ha" {
                        if value.chars().nth(value.len() - 1) == Some('/') {
                            self.apis[1].url = value;
                        } else {
                            self.apis[1].url = value + "/";
                        }
                        break;
                    }
                }
            }
        }

        // for api in self.apis {
        // }
        // self.api.version = String::from(API_VERSION);
    }
}
